chat-client.ts

```diff
@@ -0,0 +1,144 @@
+import memoizeOne from "memoize-one";
+
+import { ClientOptions } from "./types";
+
+const initializeWSClient = memoizeOne(async (projectId: string) => {
+  return new Promise<WebSocket>((resolve, reject) => {
+    const wsClient = new WebSocket("wss://api.commonbase.com/chats");
+
+    function cleanup() {
+      wsClient.removeEventListener("error", handleError);
+      wsClient.removeEventListener("message", handleMessage);
+    }
+
+    const handleError = (error: Event) => {
+      reject(error);
+      cleanup();
+    };
+    wsClient.addEventListener("error", handleError);
+
+    const handleMessage = (event: MessageEvent<string>): void => {
+      const msg = JSON.parse(event.data);
+      if (msg.type === "init") {
+        if (msg.error) {
+          reject(msg.error);
+          cleanup();
+          return;
+        }
+        if (msg.initialized) {
+          if (msg.error) {
+            reject(msg.error);
+          } else {
+            resolve(wsClient);
+          }
+          cleanup();
+        }
+      }
+    };
+    wsClient.addEventListener("message", handleMessage);
+
+    function handleOpen() {
+      wsClient.send(JSON.stringify({ type: "init", projectId }));
+      wsClient.removeEventListener("open", handleOpen);
+    }
+    wsClient.addEventListener("open", handleOpen);
+  });
+});
+
+function calculateExponentialBackoffMs(attempt: number): number {
+  const INTERVAL_MS = 100;
+  const EXPONENTIAL = 1;
+  return INTERVAL_MS * EXPONENTIAL ** attempt;
+}
+
+class _StreamingChatResponse {
+  isCompleted = false;
+  chunks: string[] = [];
+
+  constructor(
+    startStream: (events: {
+      onChunk: (chunk: string) => void;
+      onCompleted: () => void;
+    }) => void,
+  ) {
+    startStream({
+      onChunk: (chunk) => {
+        this.chunks.push(chunk);
+        this.listeners.chunk.forEach((f) => f());
+      },
+      onCompleted: () => {
+        this.isCompleted = true;
+        this.listeners.completed.forEach((f) => f());
+      },
+    });
+  }
+
+  private listeners = Object.fromEntries(
+    (["chunk", "completed"] as const).map((key) => [key, new Set()]),
+  ) as Record<"chunk" | "completed", Set<() => void>>;
+  on(eventName: keyof typeof this.listeners, listener: () => void) {
+    this.listeners[eventName].add(listener);
+    return () => {
+      this.listeners[eventName].delete(listener);
+    };
+  }
+}
+export type StreamingChatResponse = InstanceType<typeof _StreamingChatResponse>;
+
+export class ChatClient {
+  private options: ClientOptions;
+  constructor(options: ClientOptions = {}) {
+    this.options = options || {};
```

Comment:
Why is the webSocketClient not a private member of the class?

Reply:
It's a bit unorthodox, but I wanted to make sure concurrent calls get the same websocket client, even when it's still initalizing. Keeping a promise around didn't feel right, so I went with this memoization-based approach, which then happened to also alleviate the need for a private member.

Reply:
got it, but this also means there can only be a single instance of `ChatClient`

Reply:
In general, I would lean toward allowing to have multiple `ChatClient` instances (each with its own session).

Reply:
Very true, very fair point, I'll address it in the next batch